src/decode/stream.rs

```diff
@@ -0,0 +1,497 @@
+    pub fn get_ref(&self) -> Option<&W> {
+        match &self.state {
+            Some(State::Init(output)) => Some(&output),
+            Some(State::Run(state)) => Some(state.decoder.output.get_output()),
+            None => None,
+        }
+    }
+
+    /// Get a mutable reference to the output sink
+    pub fn get_mut(&mut self) -> Option<&mut W> {
+        match &mut self.state {
+            Some(State::Init(output)) => Some(output),
+            Some(State::Run(state)) => Some(state.decoder.output.get_output_mut()),
+            None => None,
+        }
+    }
+
+    /// Consumes the stream and returns the output sink. This also makes sure
+    /// we have properly reached the end of the stream.
+    pub fn finish(mut self) -> crate::error::Result<W> {
+        if let Some(state) = self.state.take() {
+            match state {
+                State::Init(output) => {
+                    if self.tmp_len > 0 {
+                        Err(Error::LZMAError("failed to read header".to_string()))
+                    } else {
+                        Ok(output)
+                    }
+                }
+                State::Run(mut state) => {
+                    if !self.options.allow_incomplete {
+                        // Process one last time with empty input to force end of
+                        // stream checks
+                        let mut stream = Cursor::new(&self.tmp[0..self.tmp_len]);
+                        let mut range_decoder =
+                            RangeDecoder::from_parts(&mut stream, state.range, state.code);
+                        state.decoder.process(&mut range_decoder)?;
+                    }
+                    let output = state.decoder.output.finish()?;
+                    Ok(output)
+                }
+            }
+        } else {
+            // this will occur if a call to `write()` fails
+            Err(Error::LZMAError(
+                "can't finish stream because of previous write error".to_string(),
+            ))
+        }
+    }
+
+    /// Attempts to read the header and transition into a running state.
+    ///
+    /// This function will consume the state, returning the next state on both
+    /// error and success.
+    fn read_header<R: BufRead>(
+        output: W,
+        mut input: &mut R,
+        options: &Options,
+    ) -> Result<State<W>, StreamStateError<W>> {
+        let params = match LZMAParams::read_header(&mut input, options) {
+            Ok(params) => params,
+            Err(e) => {
+                return Err(StreamStateError::new(Some(State::Init(output)), e));
+            }
+        };
+
+        let len = match input.fill_buf() {
+            Ok(val) => val,
+            Err(_) => {
+                return Err(StreamStateError::new(
+                    Some(State::Init(output)),
+                    Error::LZMAError("need more input".to_string()),
+                ));
+            }
+        }
+        .len();
+
+        if len < START_BYTES {
```

Comment:
Is it an error? As far as I'm aware, `fill_buf` can return as few as 1 byte at a time, but it doesn't mean that `START_BYTES` aren't available afterwards.

From the documentation (https://doc.rust-lang.org/std/io/trait.BufRead.html#tymethod.fill_buf).

> An empty buffer returned indicates that the stream has reached EOF.

It would actually be nice to have some unit tests for this streaming API with some `BufRead` implementation that returns 1 byte at a time. Although that would be inefficient, that would also test the limits of the implementation.

Reply:
There were a couple of issues with this function that I revisited. Mainly, we need to distinguish between a fatal error and "need more bytes". That's why I was doing a length check before continuing. However, this length check should have been done before calling `read_header()`. Doing so was not possible because the compiler reasonably complains that `output` is being moved multiple times.

Instead of doing a length check, I added an `Error::HeaderTooShort` (one step towards #6 maybe) so that we can distinguish between errors and handle the "need more bytes" case.

I also added `into_output()` to get back the output sink that was moved into the `DecoderState`'s `LZBuffer`.

Lastly, I added a test case `test_bad_header` to show that we properly fail on an invalid header.

Reply:
Regarding the unit tests, I do have a couple of tests in `stream.rs` to test resilience like this. `test_stream_chunked` will test every single chunk size from 1 on the whole input size.